File: wafel/ui/input_text_with_error.py
from typing import *

import wafel.imgui as ig
from wafel.util import *
from wafel.local_state import use_state


T = TypeVar('T')


def render_input_text_with_error(
  id: str,
  value: str,
  buffer_size: int,
  width: int,
  validate: Callable[[str], T],
) -> Maybe[T]:
  ig.push_id(id)

  cursor_pos = ig.get_cursor_pos()
  cursor_pos = (
    ig.get_window_position()[0] + cursor_pos[0],
    ig.get_window_position()[1] + cursor_pos[1] - ig.get_scroll_y(),
  )

  ig.push_item_width(width)
  changed, new_value = ig.input_text('##input', value, buffer_size)
  ig.pop_item_width()

  result_value = None
  if changed:
    try:
      result_value = Just(validate(new_value))
    except:
      # TODO: Show error message
      dl = ig.get_window_draw_list()
      dl.add_rect(
        cursor_pos[0],
        cursor_pos[1],
        cursor_pos[0] + width,
        cursor_pos[1] + ig.get_text_line_height() + 2 * ig.get_style().frame_padding[1],
        ig.get_color_u32_rgba(1, 0, 0, 1),
      )
      new_value = value

  ig.pop_id()
  return result_value

File: wafel_app/src/input_text_with_error.rs
use std::fmt::Write;

use imgui::{self as ig, im_str};

// TODO: Show error message (on hover?)

pub(crate) fn render_input_text_with_error<T>(
    ui: &ig::Ui<'_>,
    id: &str,
    value: &str,
    buffer_size: usize,
    width: f32,
    validate: impl Fn(&str) -> Option<T>,
) -> Option<T> {
    let id_token = ui.push_id(id);

    let mut cursor_pos = ui.cursor_pos();
    cursor_pos[0] += ui.window_pos()[0];
    cursor_pos[1] += ui.window_pos()[1] - ui.scroll_y();

    ui.set_next_item_width(width);

    let mut buffer = ig::ImString::with_capacity(buffer_size);
    buffer.write_str(&value[..buffer_size]).unwrap();
    let changed = ui.input_text(im_str!("##input"), &mut buffer).build();

    let mut result_value = None;
    if changed {
        match validate(buffer.to_str()) {
            Some(new_value) => result_value = Some(new_value),
            None => {
                let dl = ui.get_window_draw_list();
                dl.add_rect(
                    cursor_pos,
                    [
                        cursor_pos[0] + width,
                        cursor_pos[1]
                            + ui.text_line_height()
                            + 2.0 * ui.clone_style().frame_padding[1],
                    ],
                    ig::ImColor32::from_rgb_f32s(1.0, 0.0, 0.0),
                )
                .build();
            }
        }
    }

    id_token.pop(ui);

    result_value
}
